### ISSUE TITLE
Error returning default value for a nullable type from a generic method.

### ISSUE THREAD

**Issue by scott-parkhill** [dotnet/roslyn#64994]
**Version Used**: 
C# 10

**Steps to Reproduce**:

1. Create a generic class `C<T>`
2. Create a method in the generic class that returns `T?`
3. Return `default` from that class

[Sharplab Repo](https://gist.github.com/scott-parkhill/04739688a584f65bfa45a4799ebf7d3b)

[Also another showing how default of `int?` really should be null, not 0.](https://sharplab.io/#gist:35229d119f783c91a2a29c38b9bc76ab)

[Another showing return type of `T?` actually returns `int` and not `int?`.](https://sharplab.io/#gist:c54cd8ef7fb12a3e325dbc9f69dc7d9f)

**Expected Behavior**:

Expected behaviour is for `default(T?)` to return null. In the example using `int` as the type of `T`, then trying to return `default` should return `default(T?)` which should be null (or rather, `Nullable<int>` with `HasValue` equal to `false`). 

**Actual Behavior**:

Instead, it returns 0. With further testing, it actually returns `Int32`, not `Nullable<int>`. So something is very off here. I'm indicating the return type is a nullable generic, but a nullable is not returned.

**Comment by jcouv** [comment]
The meaning of `?` is different when it is on a value type (such as `int`) or not (such as `string` or unconstrained `T`).
In those latter scenarios, the type is really still just `T` (not `System.Nulable<T>`) and the `?` is a nullable annotation (it doesn't change the type).
So when substituting `int` into `T?` you just get type `int`.

In contrast, the following code uses a constrained `T`, so `T?` means `System.Nulable<T>` and `default` therefore results in a `null`:
```
System.Console.Write(M<int>() is null); // Prints True

static T? M<T>() where T : struct 
    => default;
```

**Comment by scott-parkhill** [comment]
This doesn't really solve the issue @jcouv. When you specify a return type of `T?` and `T` is `int`, you don't expect the return type to be `int`, you expect it to be `Nullable<T>`. I don't see the difference between this and being able to do `public int Test() => "I am a string";`, it's equally misleading.

**Comment by jcouv** [comment]
> you expect it to be `Nullable<T>`

That'd be wrong. `?` has different meanings in different sorts of types. 
I agree it's harder to understand, but that's the only way we could make nullable value types (C# 2) and nullable reference types (C# 8) to work well together. `?` has one meaning for value types (but `T` is not a value type in this case, unless you add `where T : struct`) and reference or unconstrained types.

**Comment by scott-parkhill** [comment]
I don't quite understand. You're saying it's wrong for a return type to indicate the type that is being returned? Because `T?` should represent that a nullable type is being returned, whether that be a value type or a reference type. The different behaviour is the reason for the use `default` as the return value and not `null`. When the return type is given as a nullable type it should return the appropriate type. If `T` in the class is `int` then it should box it in `Nullable<int>` with `HasValue` as `false` just as the documentation and function everywhere else works, not just arbitrarily ignore that I've set the return type as a nullable and return the non-nullable type back. I realize `?` has one meaning for value types and another of reference types, but that's what `default` is for in this generic case, is it not? I mean I literally return `default(int?)` and get `0` back from this function. That doesn't make sense at all.

I don't know. I'm going to assume you probably have 10+ years experience on me here. But it just doesn't make sense that this is the behaviour at all. Whether it returns `null` or `Nullable<T>` obviously depends on what `T` is, but this seems like a runtime thing and not a compile time thing and so shouldn't throw an error. I don't know. This is the last I'll say on it because I'm at the end of my ability to speak technically on it. All I know is this does not make sense at all, especially if you're saying for a generic type to allow nullable, I have duplicate all code, one that has `where T : struct` and one that has `where T : class`. Doesn't seem all that generic to me.

**Comment by scott-parkhill** [comment]
Sorry I feel like that came out a little brusque, it's just late and I'm tired and distracted with a lecture. No rudeness meant! I promise!

**Comment by CyrusNajmabadi** [comment]
`T?` for an unconstrained generic simply means 'can have thee default value of `T`. So for `int` that means that `T?` can have the value `default(int)` (aka 0). 

It's understood that this can be unintuitive, but it matches the semantics of the runtime. We considered a different sigil here (like `??`), but we ultimately decided we preferred the simple syntax, even if it could occasionally confuse.

**Comment by scott-parkhill** [comment]
I just don't understand why this decision would be made. It makes generics not generic. If I want something to return null generically I have to make one set of code where `where T : class` and then copy and paste everything again and change it to `where T : struct`. I don't understand how this could be the intended usage versus the runtime seeing that `T` is either a value or reference type and returning `Nullable<T>` or `null` respectively. But I won't say anymore on it. You obviously have your reasons. This just doesn't make sense at all that this is the functionality.

**Comment by CyrusNajmabadi** [comment]
> I just don't understand why this decision would be made.

Because those have been the semantics of the .net runtime for almost 20 years now.  When generics were added, there wasn't even a concept of annotated nullable reference types (which only got added a few years ago) :)

> I don't understand how this could be the intended usage versus the runtime seeing that T is either a value or reference type and returning Nullable<T> or null respectively.

That would be a breaking change as that's never been how generics has worked up to now.  Even when we added nullable reference types, we did so in a fashion that would produce no breaks in the ecosystem (which has billions of lines of code that people care about and want to keep working).  

>You obviously have your reasons. 

If this was v1, we would ideally have those semantics.  But we are practically in the position we're in now because we had a set of semantics from the earliest days of .net that defined how value types, nullable-value types, and reference types worked.  And this was the compatible way to encode nullable-reference types into that equation.

We were aware of the 'gotcha' here, and found it still the best alternative of several choices, all of which had different benefits and drawbacks.

> This just doesn't make sense at all that this is the functionality.

It makes sense if you think about compat and that that's how generics worked from the start in the runtime.  We're not in the business of relitigating this stuff as we care too much about compatibility to upend the type-system in this way.  Perhaps if there's a 'great reset' at some point, this would be redone.  But there's no need for that currently, so the choices we make are made within that greater context.
